src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs

```diff
@@ -718,5 +719,93 @@ private async ValueTask<T> CatchAndRethrowExceptionAsync<T>(Func<Task<T>> taskFu
                 throw exception;
             }
         }
+
+        public async ValueTask<IReadOnlyDictionary<Tuple<NuGetFramework, string>, IReadOnlyList<IPackageReferenceContextInfo>>> GetTransitivePackageOriginAsync(PackageIdentity transitivePackage, string projectId, CancellationToken ct)
+        {
+            /** Pseudocode
+            1. Get project restore graph 
+
+            2. Filter by packages
+
+            3. Foreach direct dependency d:
+                do DFS to look for transitive dependency
+
+                if found:
+                  Add to list
+
+            4. Return list
+            */
+            ct.ThrowIfCancellationRequested();
+
+            var singleProjectId = new[] { projectId };
+            var projectsList = await GetProjectsAsync(singleProjectId, ct);
+            var project = projectsList.FirstOrDefault();
+            var packageOrigins = new Dictionary<Tuple<NuGetFramework, string>, IReadOnlyList<IPackageReferenceContextInfo>>();
+
+            if (project != default && project is PackageReferenceProject prProject)
```

Comment:
Maybe just `project != null`is better considering we do that way in other places.

Reply:
Fixed.